### ISSUE TITLE
Missed optimization: is_digit()/iter().all() (?)

### ISSUE THREAD

**Issue by SoniEx2** [rust-lang/rust#47523]
I was playing with the following variants of the same code:

```rust
// UGLY:
impl<'a> IrcCommand<'a> {
    pub fn new(cmd: &'a [u8]) -> IrcCommand<'a> {
        if cmd.len() == 3 && match (cmd[0], cmd[1], cmd[2]) {
            (b'0'...b'9', b'0'...b'9', b'0'...b'9') => {
                // TODO switch to TryFrom/TryInto once those are stable. (rust-lang/rust#33417)
                return IrcCommand::Numeric(Numeric::new(array_ref![cmd,0,3]));
            },
            _ => false
        } {
            unreachable!()
        } else {
            IrcCommand::Stringy(Stringy(cmd))
        }
    }
}

// UGLY:
impl<'a> IrcCommand<'a> {
    pub fn new(cmd: &'a [u8]) -> IrcCommand<'a> {
        if cmd.len() == 3 && match (cmd[0], cmd[1], cmd[2]) {
            (b'0'...b'9', b'0'...b'9', b'0'...b'9') => true,
            _ => false
        } {
            // TODO switch to TryFrom/TryInto once those are stable. (rust-lang/rust#33417)
            IrcCommand::Numeric(Numeric::new(array_ref![cmd,0,3]))
        } else {
            IrcCommand::Stringy(Stringy(cmd))
        }
    }
}

// UGLY:
impl<'a> IrcCommand<'a> {
    pub fn new(cmd: &'a [u8]) -> IrcCommand<'a> {
        if cmd.len() == 3 {
            match (cmd[0], cmd[1], cmd[2]) {
                (b'0'...b'9', b'0'...b'9', b'0'...b'9') => {
                    // TODO switch to TryFrom/TryInto once those are stable. (rust-lang/rust#33417)
                    IrcCommand::Numeric(Numeric::new(array_ref![cmd,0,3]))
                },
                _ => IrcCommand::Stringy(Stringy(cmd))
            }
        } else {
            IrcCommand::Stringy(Stringy(cmd))
        }
    }
}

// UGLY:
impl<'a> IrcCommand<'a> {
    pub fn new(cmd: &'a [u8]) -> IrcCommand<'a> {
        match (cmd.get(0), cmd.get(1), cmd.get(2)) {
            (Some(&b'0'...b'9'), Some(&b'0'...b'9'), Some(&b'0'...b'9')) if cmd.len() == 3 => {
                // TODO switch to TryFrom/TryInto once those are stable. (rust-lang/rust#33417)
                IrcCommand::Numeric(Numeric::new(array_ref![cmd,0,3]))
            },
            _ => IrcCommand::Stringy(Stringy(cmd))
        }
    }
}

// less ugly:
impl<'a> IrcCommand<'a> {
    pub fn new(cmd: &'a [u8]) -> IrcCommand<'a> {
        if cmd.len() == 3 && cmd.iter().all(|&b| (b as char).is_digit(10)) {
            IrcCommand::Numeric(Numeric::new(array_ref![cmd,0,3]))
        } else {
            IrcCommand::Stringy(Stringy(cmd))
        }
    }
}
```

When testing the last one (PS: I haven't tested all of them), I noticed the generated code to be much different compared to the first two cases (the second of which could be replaced with the `matches` crate, which is what I did).

Case A: (matches crate)

```asm
_ZN4uirc7command10IrcCommand3new17hca225b68bb9fc6afE:                                                                               
        .cfi_startproc                                                                                                              
        cmpq    $3, %rdx                                                                                                            
        jne     .LBB14_1                                                                                                            
        movb    2(%rsi), %r8b                                                                                                       
        addb    $-48, %r8b                                                                                                          
        cmpb    $9, %r8b                                                                                                            
        ja      .LBB14_1                                                                                                            
        movb    (%rsi), %cl                                                                                                         
        addb    $-48, %cl                                                                                                           
        cmpb    $9, %cl                                                                                                             
        ja      .LBB14_1                                                                                                            
        movb    1(%rsi), %al                                                                                                        
        addb    $-48, %al                                                                                                           
        cmpb    $9, %al                                                                                                             
        ja      .LBB14_1                                                                                                            
        movzbl  %cl, %ecx                                                                                                           
        movzbl  %al, %eax                                                                                                           
        leal    (%rcx,%rcx,4), %ecx                                                                                                 
        leal    (%rax,%rcx,2), %eax                                                                                                 
        movzbl  %r8b, %ecx                                                                                                          
        leal    (%rax,%rax,4), %eax                                                                                                 
        leal    (%rcx,%rax,2), %eax                                                                                                 
        movq    $1, (%rdi)                                                                                                          
        movw    %ax, 8(%rdi)                                                                                                        
        movq    %rsi, %rdx                                                                                                          
        jmp     .LBB14_6                                                                                                            
.LBB14_1:                                                                                                                           
        movq    $0, (%rdi)                                                                                                          
        movq    %rsi, 8(%rdi)                                                                                                       
.LBB14_6:                                                                                                                           
        movq    %rdx, 16(%rdi)                                                                                                      
        movq    %rdi, %rax                                                                                                          
        retq                                                                                                                        
.Lfunc_end14:                                                                                                                       
        .size   _ZN4uirc7command10IrcCommand3new17hca225b68bb9fc6afE, .Lfunc_end14-_ZN4uirc7command10IrcCommand3new17hca225b68bb9fc6afE
        .cfi_endproc                                                                                                                
```

Case B: (is_digit)

```asm
_ZN4uirc7command10IrcCommand3new17h6b3f6ad04dd2e514E:
	.cfi_startproc
	subq	$2, %rsp
.Lcfi43:
	.cfi_def_cfa_offset 10
	cmpq	$3, %rdx
	jne	.LBB14_1
	movzbl	(%rsi), %r8d
	leal	-48(%r8), %eax
	movl	$-48, %ecx
	cmpl	$10, %eax
	jb	.LBB14_5
	leal	-97(%r8), %eax
	movl	$-87, %ecx
	cmpl	$26, %eax
	jb	.LBB14_5
	leal	-65(%r8), %eax
	movl	$-55, %ecx
	cmpl	$25, %eax
	ja	.LBB14_7
.LBB14_5:
	addl	%r8d, %ecx
	cmpl	$10, %ecx
	jae	.LBB14_7
	movq	%rsp, %rax
	jmp	.LBB14_8
.LBB14_7:
	movb	$1, (%rsp)
	leaq	1(%rsp), %rax
.LBB14_8:
	movb	$0, (%rax)
	cmpb	$0, (%rsp)
	jne	.LBB14_9
	movzbl	1(%rsi), %r8d
	leal	-48(%r8), %eax
	movl	$-48, %ecx
	cmpl	$10, %eax
	jb	.LBB14_15
	leal	-97(%r8), %eax
	movl	$-87, %ecx
	cmpl	$26, %eax
	jb	.LBB14_15
	leal	-65(%r8), %eax
	movl	$-55, %ecx
	cmpl	$25, %eax
	ja	.LBB14_17
.LBB14_15:
	addl	%r8d, %ecx
	cmpl	$10, %ecx
	jae	.LBB14_17
	movq	%rsp, %rax
	jmp	.LBB14_18
.LBB14_17:
	movb	$1, (%rsp)
	leaq	1(%rsp), %rax
.LBB14_18:
	movb	$0, (%rax)
	cmpb	$0, (%rsp)
	jne	.LBB14_9
	movzbl	2(%rsi), %r8d
	leal	-48(%r8), %eax
	movl	$-48, %ecx
	cmpl	$10, %eax
	jb	.LBB14_22
	leal	-97(%r8), %eax
	movl	$-87, %ecx
	cmpl	$26, %eax
	jb	.LBB14_22
	leal	-65(%r8), %eax
	movl	$-55, %ecx
	cmpl	$25, %eax
	ja	.LBB14_24
.LBB14_22:
	addl	%r8d, %ecx
	cmpl	$10, %ecx
	jae	.LBB14_24
	movq	%rsp, %rax
	jmp	.LBB14_25
.LBB14_24:
	movb	$1, (%rsp)
	leaq	1(%rsp), %rax
.LBB14_25:
	movb	$0, (%rax)
	cmpb	$0, (%rsp)
	je	.LBB14_10
.LBB14_9:
	movb	1(%rsp), %al
	testb	$1, %al
	jne	.LBB14_10
.LBB14_1:
	movq	$0, (%rdi)
	movq	%rsi, 8(%rdi)
	jmp	.LBB14_11
.LBB14_10:
	movb	(%rsi), %al
	movb	1(%rsi), %cl
	addb	$-48, %al
	movzbl	%al, %eax
	addb	$-48, %cl
	movzbl	%cl, %ecx
	leal	(%rax,%rax,4), %eax
	leal	(%rcx,%rax,2), %eax
	movb	2(%rsi), %cl
	addb	$-48, %cl
	movzbl	%cl, %ecx
	leal	(%rax,%rax,4), %eax
	leal	(%rcx,%rax,2), %eax
	movq	$1, (%rdi)
	movw	%ax, 8(%rdi)
	movq	%rsi, %rdx
.LBB14_11:
	movq	%rdx, 16(%rdi)
	movq	%rdi, %rax
	addq	$2, %rsp
	retq
.Lfunc_end14:
	.size	_ZN4uirc7command10IrcCommand3new17h6b3f6ad04dd2e514E, .Lfunc_end14-_ZN4uirc7command10IrcCommand3new17h6b3f6ad04dd2e514E
	.cfi_endproc
```

**Comment by scottmcm** [comment]
The last one is flipping to unicode (`(b as char).is_digit(10)`), so I'm not surprised it's different.  Have you tried with https://doc.rust-lang.org/nightly/std/primitive.u8.html#method.is_ascii_digit instead, which I expect is more of an equivalent comparison?

**Comment by Mark-Simulacrum** [comment]
I guess I'm not sure that Rust can do much here -- I'd expect `is_digit` to be slower due to checking all of unicode instead of just ASCII digits. Is there a chance you could test the relative performance of `is_ascii_digit` by switching to nightly for just this test? I'd imagine the array access would be fairly well optimized -- it's a constant sized static data array, and should convert to be close to equivalent. I could be wrong though.

**Comment by SoniEx2** [comment]
And that's suboptimal compared to the above asm. is_digit is far more optimizable given that it uses range patterns and things.

**Comment by Mark-Simulacrum** [comment]
I don't think I understand what you're trying to say; could you rephrase? What "that" are you referring to?

**Comment by SoniEx2** [comment]
Array access is suboptimal compared to the merged range check and normalization in that asm. Just look at it:

```asm
_ZN4uirc7command10IrcCommand3new17hca225b68bb9fc6afE:
        .cfi_startproc
        cmpq    $3, %rdx                /* is slice size = 3 */
        jne     .LBB14_1                /* if not, branch to << not a numeric >> */
        movb    2(%rsi), %r8b           /* put slice[2] in r8b */
        addb    $-48, %r8b              /* it's a value in b'0' to b'9', normalize to 0u8 to 9u8, smaller values overflow into very large values but we can use this property next */
        cmpb    $9, %r8b                /* since our b'0' to b'9' has been turned into 0u8 to 9u8, we can just compare for values between 0 and 9 */
        ja      .LBB14_1                /* and branch if it's not within range */
        movb    (%rsi), %cl             /* repeat the same for slice[0] */
        addb    $-48, %cl
        cmpb    $9, %cl
        ja      .LBB14_1
        movb    1(%rsi), %al            /* repeat it for slice[1] */
        addb    $-48, %al
        cmpb    $9, %al
        ja      .LBB14_1
        movzbl  %cl, %ecx               /* moving things around, zero-extending */
        movzbl  %al, %eax
        leal    (%rcx,%rcx,4), %ecx     /* this is actually a multiply by 5 */
        leal    (%rax,%rcx,2), %eax     /* and this is a multiply by 2 with an add */
        movzbl  %r8b, %ecx
        leal    (%rax,%rax,4), %eax     /* multiply by 5 */
        leal    (%rcx,%rax,2), %eax     /* multiply by 2 with an add */
        movq    $1, (%rdi)              /* set the variant (IrcCommand::Numeric) */
        movw    %ax, 8(%rdi)            /* Numeric the struct is Numeric<'a>(u16, &'a [u8; 3]). load up the u16 part of it */
        movq    %rsi, %rdx              /* shuffle things around so we can reuse the code below */
        jmp     .LBB14_6                /* go to << common path >> */
.LBB14_1:                               /* << not a numeric >> */
        movq    $0, (%rdi)              /* set the variant (IrcCommand::Stringy) */
        movq    %rsi, 8(%rdi)           /* &'a [u8] is a fat pointer, load it's contents. */
.LBB14_6:                               /* << common path >> */
        movq    %rdx, 16(%rdi)          /* load the contents (for Numeric) or length (for Stringy) */
        movq    %rdi, %rax              /* put result in "return slot" */
        retq                            /* return */
.Lfunc_end14:
        .size   _ZN4uirc7command10IrcCommand3new17hca225b68bb9fc6afE, .Lfunc_end14-_ZN4uirc7command10IrcCommand3new17hca225b68bb9fc6afE
        .cfi_endproc
```

**Comment by Mark-Simulacrum** [comment]
So you're saying that `is_ascii_digit` could be made faster? That's likely true. It's possible we'd take a PR for it, but to an extent, I'm not sure that such changes that will make the code readability harder are likely to be made.

**Comment by SoniEx2** [comment]
Yes: matches!(u8, b'0'...b'9') with the matches crate.

**Comment by Mark-Simulacrum** [comment]
A PR seems potentially acceptable, though I'd prefer that the other code within the `is_ascii_*` methods would also be updated to use similar checks.

**Comment by SoniEx2** [comment]
So make the compiler able to look into the table and convert it into that thing? Then every code like that can be optimized. And in the future we might be able to optimize the whole lookup table out! Think about the implications that would have for encoding_rs, for example.

**Comment by Mark-Simulacrum** [comment]
Well, it's possible that the compiler (or, rather, LLVM) could be modified to make that happen, but I suspect that changing the code is a better use of everyone's time. I suspect that it is not always true that lookup tables will be slower than this range comparison code.

**Comment by LingMan** [comment]
Narrator: "Three years later..."

TL;DR: Looks like this is fixed.

Since the examples given don't compile as is (what's an IrcCommand?) and this was apparently about the branch condition anyway, I've reduced the example to an Fn returning a bool:
```rust
pub fn foo(cmd: &[u8]) -> bool {
    cmd.len() == 3 && cmd.iter().all(|&b| (b as char).is_digit(10))
}

pub fn bar(cmd: &[u8]) -> bool {
    cmd.len() == 3 && cmd.iter().all(|&b| (b as char).is_ascii_digit())
}
```
At the time 1.23 was the latest stable version and presumably tested against. `is_ascii_digit` was still unstable though. Luckily it was stabilized in 1.24 just a few weeks later, so that's what I'll compare to.

With 1.24 `foo` spat out 112 lines of assembly, while `bar` weighted in at 55 lines. Nowadays with 1.52 both compile to the exact same assembly with just 19 lines:
```asm
example::bar:
        cmp     rsi, 3
        jne     .LBB0_4
        movzx   eax, byte ptr [rdi]
        add     eax, -48
        cmp     eax, 10
        jae     .LBB0_4
        movzx   eax, byte ptr [rdi + 1]
        add     eax, -48
        cmp     eax, 9
        ja      .LBB0_4
        movzx   eax, byte ptr [rdi + 2]
        add     eax, -48
        cmp     eax, 10
        setb    al
        ret
.LBB0_4:
        xor     eax, eax
        ret
```
For details: https://godbolt.org/z/rnqxorz8f

**Comment by SoniEx2** [comment]
Huh, nice! Wonder if it had something to do with for_each/etc, we've heard those have been huge for performance. Thanks for the update.

**Comment by scottmcm** [comment]
One hypothesis: Any time I see something about `len` and slice iterators, I think it might be thanks to https://github.com/rust-lang/rust/pull/52206/files#diff-713b9b67c5c17e062aece3c7b7981da9d51047b0280ae94e6161ec2711865301R2354
